float64 px
float64 py
float64 pz
float64 ox
float64 oy
float64 oz
float64 ow